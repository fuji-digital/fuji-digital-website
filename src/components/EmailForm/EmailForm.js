import { useState } from "react";
import { Formik } from "formik";
import emailjs from "@emailjs/browser";
import * as Yup from "yup";
import { motion } from "framer-motion";

import errorIcon from "../../images/icons/error-icon.svg";

import {
  StyledModal,
  StyledInputBox,
  StyledButton,
  StyledButtonContainer,
  StyledForm,
  StyledFujit,
  StyledError,
  StyledSubmitError,
  ErrorIconImage,
} from "./EmailForm.styles";

import pinkPentagonFujit from "../../images/fujits/pink-pentagon-fujit.svg";

export const EmailForm = () => {
  const [isSubmitted, setIsSubmitted] = useState(false);
  const [errorOnSubmission, setErrorOnSubmission] = useState(false);
  const [formDetails, setFormDetails] = useState({
    name: "",
    email: "",
    description: "",
  });

  const sendEmail = () => {
    emailjs
      .send(
        process.env.REACT_APP_EMAILJS_SERVICE_ID,
        process.env.REACT_APP_EMAILJS_TEMPLATE_ID,
        formDetails,
        {
          publicKey: process.env.REACT_APP_EMAILJS_PUBLIC_KEY,
        }
      )
      .then(
        () => {
          console.log("Email Submitted");
          setIsSubmitted(true);
        },
        (error) => {
          console.log("Email Failed", error);
          setErrorOnSubmission(true);
        }
      );
  };

  const formRequirements = Yup.object().shape({
    name: Yup.string()
      .min(2, `Please enter in your full name`)
      .max(50, `Please do not exceed 50 characters`)
      .required(`Required`),
    email: Yup.string()
      .matches(
        /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
        "Please enter in a valid email address"
      )
      .min(2, "Please enter in a valid email address")
      .max(50, "Please do not exceed 50 characters")
      .required("Required"),
    description: Yup.string()
      .max(500, "Please do not exceed 500 characters")
      .required("Required"),
  });

  return (
    <StyledModal>
      <h3>Contact Us</h3>
      {!errorOnSubmission && (
        <StyledFujit src={pinkPentagonFujit} alt="A Pink mascot waving" />
      )}
      {isSubmitted && !errorOnSubmission ? (
        <p>
          Thank you for submitting your question, our Fujits are working on an
          answer right away ðŸ˜Š
        </p>
      ) : (
        <Formik
          initialValues={{ name: ``, email: ``, description: `` }}
          validationSchema={formRequirements}
          validator={() => ({})}
          onSubmit={() => {
            sendEmail();
          }}
        >
          {({
            values,
            errors,
            touched,
            handleChange,
            handleBlur,
            handleSubmit,
            isSubmitting,
          }) => (
            <>
              {errorOnSubmission && (
                <StyledSubmitError>
                  <ErrorIconImage src={errorIcon}></ErrorIconImage>
                  <div data-testid="submission-error-message">
                    {" "}
                    Error has occured, please try again later or contact us
                    directly by clicking on{" "}
                    <a
                      href={`mailto:info@fujidigital.co.uk?subject=Submission&body=Hello, ${values.description} Thank you, ${values.name}`}
                    >
                      info@fujidigital.co.uk
                    </a>{" "}
                    (this will open a new email window)
                  </div>
                </StyledSubmitError>
              )}
              <StyledForm
                onSubmit={(event) => {
                  handleSubmit(event);
                  setFormDetails(values);
                  event.preventDefault();
                }}
              >
                <StyledInputBox
                  inputwarning={errors.name && touched.name && errors.name}
                >
                  <label id="name-label">Name</label>
                  <input
                    type="text"
                    id="name"
                    placeholder="Enter name here..."
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.name}
                    aria-describedby="name-label"
                  ></input>
                  <StyledError data-testid="name-error-message">
                    {errors.name && touched.name && errors.name}
                  </StyledError>
                </StyledInputBox>

                <StyledInputBox
                  inputwarning={errors.email && touched.email && errors.email}
                >
                  <label id="email-label">Email</label>
                  <input
                    type="email"
                    id="email"
                    placeholder="Enter email here..."
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.email}
                    aria-describedby="email-label"
                  ></input>
                  <StyledError data-testid="email-error-message">
                    {errors.email && touched.email && errors.email}
                  </StyledError>
                </StyledInputBox>

                <StyledInputBox
                  inputwarning={
                    errors.description &&
                    touched.description &&
                    errors.description
                  }
                >
                  <label id="description-label">How can we help?</label>
                  <textarea
                    type="textarea"
                    id="description"
                    placeholder="Enter question here..."
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.description}
                    aria-describedby="description-label"
                  ></textarea>
                  <StyledError data-testid="description-error-message">
                    {errors.description &&
                      touched.description &&
                      errors.description}
                  </StyledError>
                </StyledInputBox>
                <motion.div
                  whileHover={{ scale: 1.03 }}
                  whileTap={{ scale: 1.01 }}
                >
                  <StyledButtonContainer>
                    <StyledButton
                      type="submit"
                      data-testid="submit-button"
                      disabled={isSubmitting && !errorOnSubmission}
                    >
                      Submit
                    </StyledButton>
                  </StyledButtonContainer>
                </motion.div>
              </StyledForm>
            </>
          )}
        </Formik>
      )}
    </StyledModal>
  );
};
