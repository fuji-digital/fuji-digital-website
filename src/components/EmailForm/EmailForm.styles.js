import styled from "styled-components";

export const StyledModal = styled.div``;

export const StyledInputBox = styled.div`
  display: flex;
  flex-direction: column;
  color: white;
  textarea {
    box-sizing: border-box;
    resize: vertical;
    min-height: 7rem;
  }
`;
