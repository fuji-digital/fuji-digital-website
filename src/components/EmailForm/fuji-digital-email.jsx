import * as React from "react";
import { Html, Button } from "@react-email/components";

export const FujiDigitalEmail = () => {
  return (
    <Html lang="en">
      <h1>
        Thank you for your question, we will get back to you as soon as
        possible!
      </h1>
    </Html>
  );
};
