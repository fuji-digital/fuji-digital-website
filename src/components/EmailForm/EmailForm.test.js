import renderer from "react-test-renderer";
import { render, screen, waitFor } from "@testing-library/react";
import { EmailForm } from "./EmailForm";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";

jest.mock("@emailjs/browser", () => ({
  send: jest.fn(() => Promise.resolve({ status: 200, text: "OK" })),
}));

describe("EmailForm", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("matches snapshot", () => {
    const component = renderer.create(<EmailForm />).toJSON();
    expect(component).toMatchSnapshot();
  });

  it("loads thank you message when submitted", async () => {
    require("@emailjs/browser").send.mockImplementation(() =>
      Promise.resolve({ status: 200, text: "OK" })
    );

    const user = userEvent.setup();
    render(<EmailForm />);

    await act(async () => {
      await user.type(
        screen.getByRole("textbox", { description: /Name/i }),
        "John"
      );
      await user.type(
        screen.getByRole("textbox", { description: /Email/i }),
        "email@email.com"
      );
      await user.type(
        screen.getByRole("textbox", { description: /How can we help?/i }),
        "description"
      );

      await user.click(screen.getByRole("button", { name: /Submit/i }));
    });

    await waitFor(() =>
      expect(
        screen.getByText(
          "Thank you for submitting your question, our Fujits are working on an answer right away ğŸ˜Š"
        )
      ).toBeInTheDocument()
    );
  });

  it("checks if the input is correct and errors appear if not", async () => {
    const user = userEvent.setup();
    render(<EmailForm />);

    await act(async () => {
      const longDescriptionText = "g".repeat(501);

      await user.type(
        screen.getByRole("textbox", { description: /Name/i }),
        "a"
      );

      await user.type(
        screen.getByRole("textbox", { description: /Email/i }),
        "emailemail.com"
      );

      await user.type(
        screen.getByRole("textbox", { description: /How can we help?/i }),
        longDescriptionText
      );

      await user.click(screen.getByRole("button", { name: /submit/i }));
    });

    const nameValidationErrors = await screen.findByTestId(
      "name-error-message"
    );
    const emailValidationErrors = await screen.findByTestId(
      "email-error-message"
    );
    const descriptionValidationErrors = await screen.findByTestId(
      "description-error-message"
    );

    await waitFor(async () =>
      expect(nameValidationErrors).toHaveTextContent(
        "Please enter in your full name"
      )
    );
    await waitFor(async () =>
      expect(emailValidationErrors).toHaveTextContent(
        "Please enter in a valid email address"
      )
    );
    await waitFor(async () =>
      expect(descriptionValidationErrors).toHaveTextContent(
        "Please do not exceed 500 characters"
      )
    );

    await act(async () => {
      const longNameText = "g".repeat(51);

      await user.type(
        screen.getByRole("textbox", { description: /Name/i }),
        longNameText
      );
    });

    await waitFor(async () =>
      expect(nameValidationErrors).toHaveTextContent(
        "Please do not exceed 50 characters"
      )
    );
  }, 100000);

  it("checks if the input has been entered into", async () => {
    const user = userEvent.setup();
    render(<EmailForm />);

    await act(async () => {
      await user.click(screen.getByRole("button", { name: /submit/i }));
    });

    const nameValidationErrors = await screen.findByTestId(
      "name-error-message"
    );
    const emailValidationErrors = await screen.findByTestId(
      "email-error-message"
    );
    const descriptionValidationErrors = await screen.findByTestId(
      "description-error-message"
    );

    await waitFor(async () =>
      expect(nameValidationErrors).toHaveTextContent("Required")
    );
    await waitFor(async () =>
      expect(emailValidationErrors).toHaveTextContent("Required")
    );
    await waitFor(async () =>
      expect(descriptionValidationErrors).toHaveTextContent("Required")
    );
  });

  it("loads error message when submitted", async () => {
    require("@emailjs/browser").send.mockImplementation(() =>
      Promise.reject({ text: "FAILED" })
    );

    const user = userEvent.setup();
    render(<EmailForm />);

    await act(async () => {
      await user.type(
        screen.getByRole("textbox", { description: /Name/i }),
        "John"
      );
      await user.type(
        screen.getByRole("textbox", { description: /Email/i }),
        "email@email.com"
      );
      await user.type(
        screen.getByRole("textbox", { description: /How can we help?/i }),
        "description"
      );

      await user.click(screen.getByRole("button", { name: /Submit/i }));
    });

    await waitFor(() =>
      expect(
        screen.getByText(
          "Thank you for submitting your question, our Fujits are working on an answer right away ğŸ˜Š"
        )
      ).toBeInTheDocument()
    );
  });
});
