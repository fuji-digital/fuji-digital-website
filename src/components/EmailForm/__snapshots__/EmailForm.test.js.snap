// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailForm matches snapshot 1`] = `
<div
  className="sc-aYaIB hoOdjQ"
>
  <h3>
    Contact Us
  </h3>
  <img
    aria-hidden="true"
    className="sc-eqUzNf kJDyOC"
    src="pink-pentagon-fujit.svg"
  />
  <form
    className="sc-gEvDqW aVfNB"
    onSubmit={[Function]}
  >
    <div
      className="sc-fqkwJk ktUjlv"
    >
      <label
        id="name-label"
      >
        Name
      </label>
      <input
        aria-describedby="name-label"
        id="name"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter name here..."
        type="text"
        value=""
      />
      <div
        className="sc-gsFSjX exgqcy"
        data-testid="name-error-message"
      />
    </div>
    <div
      className="sc-fqkwJk ktUjlv"
    >
      <label
        id="email-label"
      >
        Email
      </label>
      <input
        aria-describedby="email-label"
        id="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter email here..."
        type="email"
        value=""
      />
      <div
        className="sc-gsFSjX exgqcy"
        data-testid="email-error-message"
      />
    </div>
    <div
      className="sc-fqkwJk ktUjlv"
    >
      <label
        id="description-label"
      >
        How can we help?
      </label>
      <textarea
        aria-describedby="description-label"
        id="description"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter question here..."
        type="textarea"
        value=""
      />
      <div
        className="sc-gsFSjX exgqcy"
        data-testid="description-error-message"
      />
    </div>
    <div>
      <div
        className="sc-dcJtft faUbuY"
      >
        <button
          className="sc-iGgVNO edfwlg"
          data-testid="submit-button"
          disabled={false}
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
</div>
`;
